core/src/main/java/jenkins/tasks/SimpleBuildWrapper.java

```diff
@@ -120,13 +146,39 @@ public void setDisposer(@NonNull Disposer disposer) {
         /**
          * Attempt to clean up anything that was done in the initial setup.
          * @param build a build being run
-         * @param workspace a workspace of the build
-         * @param launcher a way to start commands
+         * @param workspace a workspace of the build; may be {@code null} unless {@link #requiresWorkspace()} returns {@code true} (the default)
+         * @param launcher a way to start commands; may be {@code null} unless {@link #requiresLauncher()} returns {@code true} (the default)
          * @param listener a way to report progress
          * @throws IOException if something fails; {@link AbortException} for user errors
          * @throws InterruptedException if tear down is interrupted
          */
-        public abstract void tearDown(Run<?,?> build, FilePath workspace, Launcher launcher, TaskListener listener) throws IOException, InterruptedException;
+        public abstract void tearDown(@NonNull Run<?,?> build, @CheckForNull FilePath workspace, @CheckForNull Launcher launcher, @NonNull TaskListener listener) throws IOException, InterruptedException;
+
+        /**
+         * Indicates whether or not this wrapper callback requires a launcher.
+         * <p>
+         * If this return {@code false}, this means that {@link #tearDown(Run, FilePath, Launcher, TaskListener)}
+         * will accept {@code null} for its {@code launcher} parameter.
+         *
+         * @return {@code true} when this wrapper callback requires a launcher; {@code false} otherwise.
+         * @since TODO
+         */
+        public boolean requiresLauncher() {
+            return true;
+        }
+
+        /**
+         * Indicates whether or not this wrapper callback requires a workspace.
+         * <p>
+         * If this return {@code false}, this means that {@link #tearDown(Run, FilePath, Launcher, TaskListener)}
+         * will accept {@code null} for its {@code workspace} parameter.
+         *
+         * @return {@code true} when this wrapper callback requires a workspace; {@code false} otherwise.
+         * @since TODO
+         */
+        public boolean requiresWorkspace() {
+            return true;
+        }
```

Comment:
Redundant. Either the wrapper setup & teardown both require a node context (~ workspace + launcher), or neither do.

Reply:
The `Disposer` class is public API, not restricted to `SimpleBuildWrapper`s; so I figured it made sense to give it support for separate configuration.

Also, without these methods, there is nothing to refer to in the tearDown() JavaDoc to explain when the argument may be null, because there is no link to the original wrapper in this class.

An org search suggests there is no plugin using `Disposer` outside a wrapper context, so if you're happy with just making the workspace/launcher nullable, then I can indeed drop these.

Reply:
Yes a disposer matches a wrapper. It is only a separate type for reasons of serialization: you must carefully manage state of the disposer.

Reply:
Methods dropped; JavaDoc adjusted.